### PR TITLE
Refactor Multi-Agent Sensors API and allow dynamic adding of Sensors

### LINKED ISSUE #856: Add clean Python API to add sensors without restarting the simulator
## 🚀 Feature

According to @erikwijmans , we can actually add sensors after the simulator has been instationated by we currently don't have clean bindings to do so. 

` You need to add the sensor spec to the agent's _sensors SensorSuite via .add(sensor_object) and then you add the actual rendering object to the simulators _sensors dictionary (the habitat_sim.simulator.Sensor object)`

We should add a nice Python function to do so.
## Motivation
Removes another reason to have to restart the simulator and is a useful feature. It would be useful to doing things like creating 360 sensors on the fly in Habitat_Lab. 

## Pitch
You should be able to add and remove sensors without restarting the environment. (A fun use case would be a robot turning on an arm mounted camera for example).

### PR DESCRIPTION
**Pull request by Skylion007**
## Motivation and Context

Closes #856 

This also allows other static sensors to be added to the scene (like security cameras / static ortho cameras / bird eye views etc...)

## How Has This Been Tested
Locally. Will add a test case shortly, just want to finalize the API. 

Also expanded the _sensor_dict to be multi-agent compatible, let me know if I should be doing it in a different way, like by making _sensors a ```dict[int, dict]``` instead. 
With this, the only thing unblocking multi-agent support on the habitat-sim part would be testing and the step function.

## Types of changes

- [X] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have completed my CLA (see **CONTRIBUTING**)
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: habitat_sim/agent/agent.py [hunk 1]
```diff
@@ -142,9 +142,10 @@ def reconfigure(
         if reconfigure_sensors:
             self._sensors.clear()
             for spec in self.agent_config.sensor_specifications:
-                self._sensors.add(
-                    hsim.CameraSensor(self.scene_node.create_child(), spec)
-                )
+                self.add_sensor(spec)
+
+    def add_sensor(self, spec: hsim.SensorSpec) -> None:
```
Comment (Skylion007) [inline habitat_sim/agent/agent.py:147]: Should the agent config be updated here for dynamically adding the Sensor?

Comment (erikwijmans) [reply habitat_sim/agent/agent.py:147]: Yeah. Also makes sense to add an assert that the sensor uuid isn't already in the config.

File: habitat_sim/simulator.py [hunk 2]
```diff
@@ -25,12 +25,17 @@
...
 from habitat_sim.logging import logger
 from habitat_sim.nav import GreedyGeodesicFollower, NavMeshSettings, PathFinder
-from habitat_sim.sensor import SensorType
+from habitat_sim.sensor import SensorSpec, SensorType
 from habitat_sim.sensors.noise_models import make_sensor_noise_model
 from habitat_sim.sim import SimulatorBackend, SimulatorConfiguration
 from habitat_sim.utils.common import quat_from_angle_axis
 
 
+def _overwrite_sensor_spec_uuid(spec: SensorSpec, agent_id: int = 0) -> SensorSpec:
```
Comment (Skylion007) [inline habitat_sim/simulator.py:34]: @erikwijmans Is overwriting the UUID like this for multi-agents the way to go?

Comment (Skylion007) [reply habitat_sim/simulator.py:34]: Maybe I should make this method a class function so that it can be overrided?

Comment (erikwijmans) [reply habitat_sim/simulator.py:34]: IMO returning a list of observation-dicts makes more sense for multi-agent

File: habitat_sim/agent/agent.py [hunk 3]
```diff
@@ -142,9 +142,18 @@ def reconfigure(
         if reconfigure_sensors:
             self._sensors.clear()
             for spec in self.agent_config.sensor_specifications:
-                self._sensors.add(
-                    hsim.CameraSensor(self.scene_node.create_child(), spec)
-                )
+                self.add_sensor(spec, modify_agent_config=False)
+
+    def add_sensor(
```
Comment (erikwijmans) [inline habitat_sim/agent/agent.py:147]: This should be a private method since it doesn't make sense for anyone to call it besides the simulator.

File: habitat_sim/simulator.py (hunk 2 above)
Comment (Skylion007) [reply habitat_sim/simulator.py:34]: @erikwijmans Yeah, I just wanted to minimize API breakage between Habitat-API and Habitat-Lab for now. This is for the use case for having multiple static sensors around the scene (IE. Security Cameras) or alternatively agents that are moved outside of step.

File: habitat_sim/agent/agent.py (hunk 3 above)
Comment (Skylion007) [reply habitat_sim/agent/agent.py:147]: Sounds good.

File: habitat_sim/simulator.py (hunk 2 above)
Comment (erikwijmans) [reply habitat_sim/simulator.py:34]: But the question was specifically "is overwriting the UUID like this for multi-agents the way to go?".

We can always still return just a dict when there is a single agent.

Comment (Skylion007) [reply habitat_sim/simulator.py:34]: While UUID implies UNIQUE to me, meaning that any sensor regardless of agent should have a UNIQUE UID.

Comment (Skylion007) [reply habitat_sim/simulator.py:34]: Since the UUID is part of the config though, it prevents a config from being reused for multiple agents so some UUID mutation has / should occur in my opinion.

Comment (Skylion007) [reply habitat_sim/simulator.py:34]: Maybe the UUID overwriting could be done in Habitat-Lab?

Comment (erikwijmans) [reply habitat_sim/simulator.py:34]: I disagree that it has to be globally unique.  It has to be unique for that agent, but I don't see what it needs to be globally unique.

**Review by jturner65 — APPROVED** [review]
LGTM.

**Comment by Skylion007** [comment]
@erikwijmans Completely redesigned the API as you asked. I still need to fix that one faulty test case in tests, but overloading our APIs like that really makes typing and API design a bit difficult.

File: habitat_sim/simulator.py (hunk 2 above)
Comment (Skylion007) [reply habitat_sim/simulator.py]: @erikwijmans I just started the redesign, but our call guarantees get quite messy but doing this. Really don't like returning different types like that...

**Comment by Skylion007** [comment]
@erikwijmans The question is should we have the user specify a dict of actions or a list? I am partial to a dict since it allow some agents to have NOOP actions, but I am open to suggestions.

File: habitat_sim/simulator.py (hunk 2 above)
Comment (Skylion007) [reply habitat_sim/simulator.py]: Okay, I think I cleaned it up quite a bit.

**Comment by Skylion007** [comment]
@erikwijmans I think we may want step to return a `Dict[str, Dict]`, instead of a List. Thoughts?

File: habitat_sim/simulator.py [hunk 4]
```diff
@@ -106,20 +106,20 @@ def __attrs_post_init__(self) -> None:
...
-        for sensor in self._sensors.values():
-            sensor.close()
-            del sensor
+        for agent_sensorsuite in self.__sensors:
+            for sensor in agent_sensorsuite.values():
+                sensor.close()
+                del sensor
 
-        self._sensors = {}
+        self.__sensors = []
```
Comment (erikwijmans) [inline habitat_sim/simulator.py:114]: Why was this changed to `__`?

**Review by erikwijmans — COMMENTED** [review]
In terms of returning a list vs. returning a dict, since the actions are input as a dict, returning a dict makes sense.


Then, for consistency, having `get_sensor_observations` also return a dictionary makes sense

File: habitat_sim/simulator.py (hunk 4 above)
Comment (Skylion007) [reply habitat_sim/simulator.py:114]: For backwards compatibility. It's not a public API, but we have a lot of calls into. I emulate the old behavior with an '@property' getter.

File: habitat_sim/simulator.py [hunk 5]
```diff
@@ -247,36 +256,96 @@ def initialize_agent(
...
+        Dict[str, Union[bool, ndarray, "Tensor"]],
+    ]:
         self._num_total_frames += 1
-        collided = self._default_agent.act(action)
-        self._last_state = self._default_agent.get_state()
+        return_single = False
+        if not isinstance(action, dict):
+            action = {self._default_agent_id: action}
+            return_single = True
+        agent_ids = sorted(action.keys())
```
Comment (Skylion007) [inline habitat_sim/simulator.py:330]: I really don't like this behavior, it's kind of wonky to return a List of Dicts in sorted order. Should really be a Dict of Dict's, right @erikwijmans ?

Comment (erikwijmans) [reply habitat_sim/simulator.py:330]: Yeah, Dict of Dict's makes sense.  `get_sensor_observations` should also then be a dict of dicts for consistency.

File: habitat_sim/simulator.py (hunk 4 above)
Comment (erikwijmans) [reply habitat_sim/simulator.py:114]: Hmm, well, we should make a public API for getting the sensor objects.

Comment (Skylion007) [reply habitat_sim/simulator.py:114]: Yeah, that can be a separate PR though.

File: habitat_sim/simulator.py (hunk 5 above)
Comment (Skylion007) [reply habitat_sim/simulator.py:330]: Less sure about this since the IDs are passed in directly as a list.

Comment (Skylion007) [reply habitat_sim/simulator.py:330]: Both solutions are annoying for different reasons. Iterating through a dict of a dicts is a bit painful and requires some subtle instancing to detect properly compared to a list at least but figuring out which ordering is return is a bit annoying.

We could always sort the list of agent dicts by agent_ids, but that can get a bit messy if not all agents generate observations...

Comment (erikwijmans) [reply habitat_sim/simulator.py:334]: Yeah, I agree that it isn't ideal.  My thought is that even if dict of dicts is slightly more awkward, consistency is better because if one is list of dicts while the other is dict of dicts, that will require two different code paths.

What about returning an OrderedDict so `.values()` get's you the current list?

**Comment by Skylion007** [comment]
@erikwijmans I don't this will actually work as is in habitat_lab, because habitat_lab currently maps the action as a dict of the action value (str or int) to a the action arguments (like the location for teleport). To pass a multi-agent action on Habitat-Lab won't work since all the action values [str, int] are wrapped in a dict and dicts are an unhashable type and therefore cannot be keys.

**Comment by Skylion007** [comment]
I don't think sets or lists are hashable types, either... Maybe we need to create a MultiAgent Action class?

**Comment by erikwijmans** [comment]
I think above got resolved in DMs?

**Comment by Skylion007** [comment]
@erikwijmans Yeah, it did. Happy with how it is now.

**Review by erikwijmans — APPROVED** [review]

File: habitat_sim/simulator.py [hunk 6]
```diff
@@ -216,17 +225,46 @@ def __set_from_config(self, config: Configuration):
...
+
+    def _update_simulator_sensors(self, uuid: str, agent_id: int = 0) -> None:
+        self.__sensors[agent_id][uuid] = Sensor(
+            sim=self, agent=self.get_agent(agent_id), sensor_id=uuid
+        )
 
-        for i in range(len(self.agents)):
-            self.initialize_agent(i)
+    def add_sensor(self, sensor_spec: SensorSpec, agent_id: int = 0) -> None:
+        if (
```
Comment (Skylion007) [inline habitat_sim/simulator.py:245]: @bigbike This is going to get messy as we add more SensorTypes. We should consider a way of making this cleaner.

**Comment by Skylion007** [comment]
So I read about into this: https://medium.com/@vermashresth/craft-and-solve-multi-agent-problems-using-rllib-and-tensorforce-a3bd1bb6f556 . and there are some current issues with the API:

* "reset should return an observation dict for all agents" . Should reset take a 'returned_agent_ids values?
* There should be a way to get reward measures for agents that don't act, (I think this is actually find depending on how we implement it on the Habitat-Lab level.)

File: habitat_sim/simulator.py (hunk 6 above)
Comment (mathfac) [reply habitat_sim/simulator.py:249]: Maybe, those checks should live close to Sensor code of particular type and should be checked before particular sensor type creation in `update_simulator_sensors` method.

File: habitat_sim/simulator.py [hunk 7]
```diff
@@ -244,43 +288,124 @@ def initialize_agent(
                 )
 
         agent.set_state(initial_state, is_initial=True)
-        self._last_state = agent.state
+        if agent_id == len(self.__last_state):
+            self.__last_state.append(agent.state)
```
Comment (mathfac) [inline habitat_sim/simulator.py:292]: Warning: mix of dictionaries and lists. What if agent_ids aren't sequential.

File: habitat_sim/simulator.py [hunk 8]
```diff
@@ -244,43 +288,124 @@ def initialize_agent(
...
+        dt: float = 1.0 / 60.0,
+    ) -> Union[
+        Dict[str, Union[bool, ndarray, "Tensor"]],
+        Dict[int, Dict[str, Union[bool, ndarray, "Tensor"]]],
+    ]:
         self._num_total_frames += 1
-        collided = self._default_agent.act(action)
-        self._last_state = self._default_agent.get_state()
+        if isinstance(action, MutableMapping):
+            return_single = False
```
Comment (mathfac) [inline habitat_sim/simulator.py:384]: nit: to be consistent with `get_observations` worth handle `return_single = True` in if condition.

File: habitat_sim/simulator.py [hunk 9]
```diff
@@ -244,43 +288,124 @@ def initialize_agent(
...
+        else:
+            action = cast(Dict[int, Union[str, int]], {self._default_agent_id: action})
+            return_single = True
+        collided_dict: Dict[int, bool] = {}
```
Comment (mathfac) [inline habitat_sim/simulator.py:388]:
```suggestion

        collided_dict: Dict[int, bool] = {}
```

File: habitat_sim/simulator.py [hunk 10]
```diff
@@ -244,43 +288,124 @@ def initialize_agent(
...
+        for agent_id, agent_act in action.items():
+            agent = self.get_agent(agent_id)
+            collided_dict[agent_id] = agent.act(agent_act)
+            self.__last_state[agent_id] = agent.get_state()
```
Comment (mathfac) [inline habitat_sim/simulator.py:392]: Worth to make `__last_state` a dict as well to avoid handling list vs dict issues, like out of index.

**Review by mathfac — CHANGES_REQUESTED** [review]
@Skylion007, thank you a lot for contributing to Multi-Agent direction. I think we should lock the scope for this PR as we will have more insights once we add first model baseline for the multi-agent functionality. Then we can make better design choices.

I think "reset should return an observation dict for all agents" make sense. For this we need to have dict or param of agent number at least.
The reward calculation is Hab-Lab area, there we can calculate measure for all agents.

Requested some changes.

File: habitat_sim/simulator.py (hunk 7 above)
Comment (Skylion007) [reply habitat_sim/simulator.py:292]: @mathfac They have to be at init time currently. There is now way to create a non-sequential agent_id

File: habitat_sim/simulator.py (hunk 8 above)
Comment (Skylion007) [reply habitat_sim/simulator.py:384]: It makes it a lot messier. I originally had it this way. See the earlier commits.

File: habitat_sim/simulator.py (hunk 10 above)
Comment (Skylion007) [reply habitat_sim/simulator.py:392]: get_agent will also currently cause out of index, so I don't see the problem?

Comment (Skylion007) [reply habitat_sim/simulator.py:392]: If we really wanted it to be clean, we would move last state handling into the agent itself. There is no reason for the Sim to keep track of it. The agent can perfectly keep track of it.

File: habitat_sim/simulator.py (hunk 6 above)
Comment (Skylion007) [reply habitat_sim/simulator.py:249]: @mathfac They should be handled in the C++ API or have a boolean on the Sensors that correspond which flags are valid / needed for an arbitrary sensor type.

File: habitat_sim/simulator.py [hunk 11]
```diff
@@ -135,7 +146,7 @@ def seed(self, new_seed: int) -> None:
         super().seed(new_seed)
         self.pathfinder.seed(new_seed)
 
-    def reset(self) -> Dict[str, ndarray]:
+    def reset(self) -> Dict[str, Union[ndarray, "Tensor"]]:
```
Comment (mathfac) [inline habitat_sim/simulator.py:149]: For now `reset()` will return observations only for agent_id == 0.

**Comment by mathfac** [comment]
Thank you for the answers regarding dict vs list. We should be good to go with the PR as soon as reset for all agents will be supported.

**Comment by unnat** [comment]
@Skylion007 First, thank you for considering me for external feedback. Also, great to see improved multi-agent support in Habitat!

With reference to:
```
        for agent_id, agent_act in action.items():
            agent = self.get_agent(agent_id)
            collided_dict[agent_id] = agent.act(agent_act)
            self.__last_state[agent_id] = agent.get_state()
```
Here, actions are being sequentially executed, which is sufficient for navigational tasks. As we tackle multi-agent tasks, we would face situations where the success of the action of one agent is dependent on the actions+states of the other agents. For example, [in this work](https://arxiv.org/pdf/1910.05512.pdf#page=8) the door opens only if another agent is close to a switch (dependence on the state of other agents). Similarly,  agents can [move heavy objects](https://arxiv.org/pdf/2007.04979.pdf#page=6) only if they are execute coordinated actions (dependence on the action of other agents).

Hence, allowing the flexibility where a user/developer can code up the environment's transition dynamics as a function of the list of actions and list of current states of the agents, might be a robust way forward for more interesting coordinated tasks.

**Comment by Skylion007** [comment]
> @Skylion007 First, thank you for considering me for external feedback. Also, great to see improved multi-agent support in Habitat!
> 
> With reference to:
> 
> ```
>         for agent_id, agent_act in action.items():
>             agent = self.get_agent(agent_id)
>             collided_dict[agent_id] = agent.act(agent_act)
>             self.__last_state[agent_id] = agent.get_state()
> ```
> 
> Here, actions are being sequentially executed, which is sufficient for navigational tasks. As we tackle multi-agent tasks, we would face situations where the success of the action of one agent is dependent on the actions+states of the other agents. For example, [in this work](https://arxiv.org/pdf/1910.05512.pdf#page=8) the door opens only if another agent is close to a switch (dependence on the state of other agents). Similarly, agents can [move heavy objects](https://arxiv.org/pdf/2007.04979.pdf#page=6) only if they are execute coordinated actions (dependence on the action of other agents).
> 
> Hence, allowing the flexibility where a user/developer can code up the environment's transition dynamics as a function of the list of actions and list of current states of the agents, might be a general way forward for interesting coordinated tasks. Perhaps, food for thought for later edits, if not now.

It seems like these conditions are best handled as Task actions in the Habitat-API though. Maybe the thing we should really allow is the ability to call step without advancing physics?.
